contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +695,93 @@ with minimal disruption to users would be:
+of a legacy in-tree plugin will get invoked. Kubelet's ProbeVolumePlugins and Kube
+Controller Manager's ProbeAttachableVolumePlugins, ProbeExpandableVolumePlugins and
+ProbeControllerVolumePlugins will call the above instead of directly probing in-tree plugins
+that will be migrated.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+pkg/volume/util/csimigration/plugins.go so that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on a common utility function:
+```
+func IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+```
+that will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of nil Plugin Object returned by FindPluginBySpec/FindPluginByName
+
+Once an in-tree plugin is no longer probed, all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return a nil plugin
+object to code paths that invoke those functions. For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return nil for a migrated
+plugin that is no longer probed in ProbeControllerVolumePlugins.
+2. FindExpandablePluginBySpec in the expand controller will return nil for a migrated
+plugin that is no longer probed in ProbeExpandableVolumePlugins.
+3. FindAttachablePluginBySpec in the attach/detach controller will return nil for
+a migrated plugin that is no longer probed in ProbeAttachableVolumePlugins.
+
+Each such code path will need to be audited and modified such that:
+1. Code preceding the invocation of Find*Plugin routines performs the necessary
+translation from in-tree to CSI so that when a Find*Plugin routine is invoked,
+it is passed a translated CSI PV rather than an in-tree PV.
+2. Upon receiving a nil plugin, the code checks migration status of a plugin based
+on spec or name rather than throwing an error.
+
+### Enhancements in Controllerss handling Persistent Volumes
+
+#### AttachDetach Controller
+When CSIMigrationInTreeOff is enabled, translation of volume specs from in-tree to
+CSI will take place at the points where Desired State of World cache gets populated
+through AddPodToVolume and ProcessPodVolumes. This avoids the down-stream
+functions at the operation generator/executor stages from having to correctly handle
+translation of volume specs for migrated PVs that are no longer probed.
+
+#### Expansion Controller
+syncHandler will be refactored to only invoke FindExpandablePluginBySpec for
```

Comment:
I'm not clear on what this paragraph means. Could you clarify?

Reply:
Overall, in this section, I call out the main changes necessary in the different controllers to work correctly with `CSIMigrationInTreeOff` feature flag enabled. 

For the Expansion Controller, I mean to say that it's `syncHandler` at https://github.com/kubernetes/kubernetes/blob/28e800245e910b65b56548f36172ce525a554dc8/pkg/controller/volume/expand/expand_controller.go#L238 will be refactored to translate the PV and determine `csiResizerName`. The current structure involving invocation of `FindExpandablePluginBySpec` followed by `volumePlugin.IsMigratedToCSI()` will no longer work with `CSIMigrationInTreeOff` feature flag enabled.